File: src/main/scala/f0/Sources.scala
package f0

trait ByteReader {
  def readByte: Byte
  def readBytes(n: Int): Array[Byte]
}

trait BitReader {
  def readBit: Boolean
}

trait Source extends ByteReader with BitReader {
  def apply[A,F](r:Reader[A,F]) = r.bind(this).get
  // TODO: these arent used, so im commenting them out - JC 11/29/11
  //def foldBytes[A](n: Int, z: A)(f: (A,Byte) => A): A
  //def state: Reg
}

object Sources {
  
  def getBit(b: Int, i: Int): Boolean = (b & i) != 0

  def fromFile(filename:String): Source = fromInputStream(new java.io.FileInputStream(filename))
  def fromProcess(p:Process): Source = fromInputStream(p.getInputStream)
  def fromArray(bytes: Array[Byte], start: Int = 0): Source = new SourceInternal {
    var byteIndex = start
    def readByte = { val b = bytes(byteIndex); byteIndex += 1; b }
    def readBytes(n: Int) = { 
      val a = new Array[Byte](n)
      Array.copy(bytes, byteIndex, a, 0, n)
      byteIndex += n
      a 
    }
// TODO: these arent used, so im commenting them out - JC 11/29/11
//    def foldBytes[A](n: Int, z: A)(f: (A,Byte) => A): A = {
//      var c = n
//      var acc = z
//      while (c > 0) { acc = f(acc, bytes(i)); i += 1; c -= 1 }
//      acc
//    }
//    val state = Reg(Map())
  }

  /**
   * reads from the stream, blocking until data is available
   * or throwing an EOFException if there is an attempt to read a byte
   * when the stream has reached EOF.
   **/
  def fromInputStream(is: java.io.InputStream): Source = new SourceInternal {
    var byteIndex = 0
    // TODO: this can probably be made more efficient,
    // because im just reading one byte at a time.
    // i could read into an array instead.
    val bytesStream = Iterator.continually({
      val i = is.read
      if(i == -1) throw new java.io.EOFException()
      i.toByte
    })
    def readByte = { val b = bytesStream.take(1).next(); byteIndex += 1; b }
    def readBytes(n: Int) = { byteIndex += n; bytesStream.take(n).toArray }
  }

  private trait SourceInternal extends Source with ByteReaderInternal with BitReaderInternal

  private trait ByteReaderInternal extends ByteReader {
    // byteIndex used to communicate with the BitReader
    // so it knows if it needs to read the next byte or not
    // TODO: i imagine theres a better way to do this.
    var byteIndex: Int
  }

  private trait BitReaderInternal extends ByteReaderInternal with BitReader {
    var bitOffset = 1
    var byteBuf = 0
    var byteBufMark = -1
    def readBit =
      if (byteBufMark != byteIndex || bitOffset == 256) {
        bitOffset = 2
        byteBuf = readByte
        byteBufMark = byteIndex
        getBit(byteBuf, 1)
      }
      else {
        val r = getBit(byteBuf,bitOffset)
        bitOffset *= 2
        r
      }
  }
}

File: fsharp/f0/f0/Sources.fs
namespace f0

open System.IO

type Source =
  abstract member ReadBit:  bool
  abstract member ReadByte: byte
  abstract member ReadBytes: int -> array<byte>
  //def foldBytes[A](n: Int, z: A)(f: (A,Byte) => A): A
  abstract member State: Reg

type Binary<'f> = 
    abstract member source: Source
    // can map over Binary[F] to change its format!

module GetBit =
    let getBit (b:int) (i:int) : bool = not ((b &&& i) = 0)

open GetBit

// TODO both ByteArraySource and InputStream source share (copy/paste)
// the same bit reading code. figure out a way to share it.
type ByteArraySource(bytes:array<byte>, start:int) as self = 
    let mutable i = start
    let mutable bitOffset = 1
    let mutable byteBuf = 0
    let mutable byteBufMark = -1
    member s.ReadByte =
        let b = bytes.[i]
        i <- i + 1
        b
    interface Source with
        member s.ReadBit: bool =
            if byteBufMark <> i || bitOffset = 256
            then
                bitOffset <- 2
                byteBuf <- int self.ReadByte
                byteBufMark <- i
                getBit byteBuf 1
            else
                let r = getBit byteBuf bitOffset
                bitOffset <- bitOffset * 2
                r
        member s.ReadByte = self.ReadByte
        member s.ReadBytes n =
            let a = Array.create n 0uy
            System.Array.Copy(bytes, i, a, 0, n)
            i <- i + n
            a 
        member s.State = Reg Map.empty

type InputStreamSource(stream:Stream) as self = 
    let mutable byteIndex = 0
    let mutable bitOffset = 1
    let mutable byteBuf = 0
    let mutable byteBufMark = -1
    member s.ReadByte =
        let i: int = stream.ReadByte()
        if i = -1 then failwith("end of stream")
        let b: byte = (byte i)
        byteIndex <- byteIndex + 1
        b
    interface Source with
        member s.ReadBit: bool =
            if byteBufMark <> byteIndex || bitOffset = 256
            then
                bitOffset <- 2
                byteBuf <- int self.ReadByte
                byteBufMark <- byteIndex
                getBit byteBuf 1
            else
                let r = getBit byteBuf bitOffset
                bitOffset <- bitOffset * 2
                r
        member s.ReadByte = self.ReadByte
        member s.ReadBytes n =
            let a = Array.create n 0uy
            for i in 0 .. (n-1) do a.[i] <- s.ReadByte
            a 
        member s.State = Reg Map.empty

module Sources =

    let FromByteArray(bytes:array<byte>): Source = new ByteArraySource(bytes, 0) :> Source
    let FromInputStream(stream:Stream): Source = new InputStreamSource(new BufferedStream(stream)) :> Source
    let FromFile(filename:string): Source = FromInputStream(File.OpenRead(filename))
    let FromStdIn() = FromInputStream(System.Console.OpenStandardInput())
